### PR TITLE
Carson Jardine Individual Assessment 2

### PR DESCRIPTION
**Pull request by carson-jardine**
Struggled with finding the total price and products in iteration 1 and 3. Also, with the hoarding part as well as the details. Not my best work, need to practice/go over this.

### CODE REVIEW

**Comment by timomitchel** [comment]
Nice job committing throughout. You have 5 commits, but in 2 hour challenges, we'd like to see around 8.

**Comment by carson-jardine** [comment]
@timomitchel Thanks Tim, I had a feeling it wasn't enough. Trying to make an effort to commit after any new test passes just need to keep practicing. :)

File: lib/product.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+class Product
+  attr_reader :category, :name, :quantity, :unit_price
+
+  def initialize(category, name, unit_price, quantity)
+    @category = category
+    @name = name
+    @quantity = quantity.to_i
+    @unit_price = unit_price.to_f.round(2)
```
Comment (timomitchel) [inline lib/product.rb:8]: No need for the `.to_f.round(2)`, this value is passed in the same way the spec asks for.

File: lib/product.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  def total_price
+    sum = @product.unit_price * @product.quantity
+  end
+
+  def hoard
+    return @is_hoarded = true
+  end
+
+  def is_hoarded?
```
Comment (timomitchel) [inline lib/product.rb:20]: This method should simply return the `@is_hoarded` instance variable.

File: lib/product.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @unit_price = unit_price.to_f.round(2)
+    @is_hoarded = false
+  end
+
+  def total_price
+    sum = @product.unit_price * @product.quantity
+  end
+
+  def hoard
+    return @is_hoarded = true
```
Comment (timomitchel) [inline lib/product.rb:17]: No need for the explicit `return` here. We aren't worried about this methods return value, but we do want to set `@is_hoarded` to true like you've done.

File: lib/product.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  def initialize(category, name, unit_price, quantity)
+    @category = category
+    @name = name
+    @quantity = quantity.to_i
+    @unit_price = unit_price.to_f.round(2)
+    @is_hoarded = false
+  end
+
+  def total_price
+    sum = @product.unit_price * @product.quantity
```
Comment (timomitchel) [inline lib/product.rb:13]: So, we don't have access to `@product` within our class. We have access to it in our test. This method should read: 
`unit_price * quantity`

File: test/product_test.rb [hunk 5]
```diff
@@ -1,44 +1,53 @@
...
   end
 
   def test_it_has_a_quantity
-    skip
+
     product = Product.new(:paper, 'toilet paper', 3.70, '10')
     assert_equal 10, product.quantity
   end
 
   def test_it_can_get_total_price
```
Comment (timomitchel) [inline test/product_test.rb:39]: I can tell some of these tests do not pass based on the implementation. Make sure to run the test suite often and let the errors guide implementation!

File: test/product_test.rb [hunk 6]
```diff
@@ -1,44 +1,53 @@
...
     product2 = Product.new(:meat, 'chicken', 4.50, '2')
 
     assert_equal 37.0, product1.total_price
     assert_equal 9.0, product2.total_price
   end
+
+  def test_is_hoarded
+    product = Product.new(:paper, 'toilet paper', 3.70, '10')
+
+    assert_equal false, product.is_hoarded?
```
Comment (timomitchel) [inline test/product_test.rb:51]: Any time we have a boolean method, we should setup and test both a `true` and `false` scenario.

File: test/shopping_cart_test.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+require 'minitest/autorun'
+require 'minitest/pride'
+require './lib/product'
+require './lib/shopping_cart'
+require 'pry'
+
+class ShoppingCartTest < Minitest::Test
+  def setup
+    @cart = ShoppingCart.new("King Soopers", "30items")
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:9]: Nice job using a setup method to DRY up your tests here.

File: test/shopping_cart_test.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+    @cart.add_product(product1)
+
+    assert_equal [product1], @cart.products
+
+
+    product2 = Product.new(:meat, 'chicken', 4.50, '2')
+    @cart.add_product(product2)
+
+    assert_equal [product1, product2], @cart.products
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:40]: This is really solid and robust testing.

File: lib/shopping_cart.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+  def details
+    cart = {
+      name: @cart.name,
+      capacity: @cart.capacity
+    }
+  end
+
+  def total_number_of_products
+    if !@products == []
```
Comment (timomitchel) [inline lib/shopping_cart.rb:23]: This `if` statement looks like it actually breaks the code. It is unnecessary. The method should read: 
```
@products.sum do |quantity|
      product.quantity
  end

File: lib/shopping_cart.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      @products.find_all do |quantity|
+      product.quantity += cart.total_number_of_products
+      end
+    end
+  end
+
+  def is_full?
+    if cart.total_number_of_products <=30
```
Comment (timomitchel) [inline lib/shopping_cart.rb:31]: On the right track here, but we don't need the `if` but rather just the condition which should read:
`total_number_of_products > capacity`

File: test/shopping_cart_test.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+    product4 = Product.new(:produce, 'apples', 0.99, '20')
+
+    @cart.add_product(product1)
+    @cart.add_product(product2)
+    @cart.add_product(product3)
+    @cart.add_product(product4)
+
+    assert_equal [product1, product3], @cart.products_by_category(:paper)
+  end
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:72]: Solid test setup and execution here.

**Comment by carson-jardine** [comment]
@timomitchel  Thanks Tim! I  went through this with my mentor the same day we did it and I definitely feel a lot more solid on the concepts and everything!

**Pull request closed without merging**